### ISSUE TITLE
User-defined forcing function signature

### ISSUE THREAD

**Issue by glwagner** [climate-machine/Oceananigans.jl#390]
Currently, user-defined forcing functions (say, `F`) have the signature

```julia
F(grid, U, Φ, i, j, k) = ...
```

per

https://github.com/climate-machine/Oceananigans.jl/blob/3ab93222717271de194ade9e9b073c1c09f571e8/src/time_steppers.jl#L135

However, I think that we should reorganize this so that the most "important" arguments are first. I think perhaps a better ordering would be 

```julia
F(i, j, k, grid, time, velocities, tracers) = ...
```

The idea is that `grid` is unlikely to be used without indices; however indices can be used without `grid`. This pattern also resembles the pattern we have developed for kernels (and thus could be used in the future to develop "differentiated" user-defined forcing functions, for example).

This also adds time as an argument, which I think is a good idea as this is needed for lots of elementary things like, eg, the tidal forcing functions desired by #212. Adding time as an argument will require some minor modifications to `calculate_interior_source_terms!`.

We should also change over all of our kernels to having `(i, j, k, grid)` first, but we don't have to address that refactoring immediately. Note that changing the forcing function signature is a significant change as it affects the user API.

**Comment by ali-ramadhan** [comment]
Would definitely be nice to be consistent. Didn't realize that `time` wasn't part of the forcing functions, that's definitely an important one.

I might vote for
```julia
F(i, j, k, t, grid, velocities, tracers) = ...
```
to mirror the way we write functions, e.g. `u(x, y, z, t)`.

**Comment by glwagner** [comment]
`time` is part of the boundary condition function signature (along with `iteration`), but it has not been included in the forcing signature yet. 

So I guess you are arguing that things like interpolation and taking derivatives of forcing functions is not important? That may be true (and I might agree) --- just want to make sure we are on the same page regarding the implications of these choices.

**Comment by ali-ramadhan** [comment]
> So I guess you are arguing that things like interpolation and taking derivatives of forcing functions is not important?

Did my comment imply this? I don't see myself needing to interpolate or differentiate forcing functions but you should be able to do it if you want/need to.

I guess I forgot the iteration number
```julia
F(i, j, k, t, iter, grid, velocities, tracers) = ...
```
although not sure when I'd use `iter` if I have `t`.

**Comment by glwagner** [comment]
> Did my comment imply this?

Yes,. your comment did imply that, because with the four arguments `i, j, k, grid` we can use the interpolation and differentiation functions we developed for turbulence closures directly on a forcing function. In other words, we could write 

```julia
Gu[i, j, k] = ... + ∂x_faa(i, j, k, grid, F.u, time, velocities, tracers)
```

Or more generally

```julia
Gu[i, j, k] = ... + mod_function(i, j, k, grid F.u, time, velocities, tracers)
```

where `mod_function` is either an interpolation or differentiation function supplied by the user somehow. 

We don't have an application for this so it may not be important, and this is a perfectly legitimate argument to make. However, it was with the motivation that this kind of shenanigans becomes possible that I proposed the signature above. 

If we alter the function signature to have arguments `(i, j, k, time, grid, ...)`, we can still do interpolation/differentiation; however it will require defining new functions for that case.

**Comment by glwagner** [comment]
> although not sure when I'd use iter if I have t

`iter` is useful for defining stochastic forcings that change every iteration (with a substep method like RK3, you must hold the value of a random variable constant over the substeps, but change on whole steps). It's not crucial right now.

**Comment by ali-ramadhan** [comment]
Hmmm, on some level I feel like differentiating, interpolating, and integrating forcing functions seems like something the user should be doing, but I could see it being useful for collecting statistics for post-processing or analysis.

If it's just a matter of switching arguments, then it sounds like a small price to pay for a potentially cool feature.

I guess at some point we'll probably want to pick a convention and refactor the operators, time-stepping kernels, and user-defined functions to all match.
